doc/aaa/RBAC_Phase2.md

```diff
@@ -0,0 +1,421 @@
+- **groupadd**: To create new groups/roles (similar to GNU [groupadd](http://man7.org/linux/man-pages/man8/groupadd.8.html))
+- **groupdel**: To delete groups/roles (similar to GNU [groupdel](http://man7.org/linux/man-pages/man8/groupdel.8.html))
+- **useredit**: To create or modify users. This method will accept a username,
+  hashed password, and a list of roles.
+
+The hamd process will add the users to the User Table in the Config DB, however,
+only the username, tenant (empty for now) and roles will be saved in the DB. The
+password will not be saved due to security considerations.
+
+##### 1.1.1.6.2 Name Service Switch
+
+In addition to providing APIs to create/modify/delete user and group (role) accounts, hamd also provides APIs to simply read user and group (role) accounts. Here's the list:
+
+- **getpwnam**: To retrieve user credentials (similar to POSIX [getpwnam](http://man7.org/linux/man-pages/man3/getpwnam.3.html))
+- **getpwuid**: To retrieve user credentials (similar to POSIX [getpwuid](http://man7.org/linux/man-pages/man3/getpwnam.3.html))
+- **getgrnam**: To retrieve group/role credentials (similar to POSIX [getgrnam](http://man7.org/linux/man-pages/man3/getgrnam.3.html))
+- **getgrgid**: To retrieve group/role credentials (similar to POSIX [getgrgid](http://man7.org/linux/man-pages/man3/getgrnam.3.html))
+
+These APIs, however, are meant to be invoked through [NSS](https://en.wikipedia.org/wiki/Name_Service_Switch) (name service switch).  That is, applications running in containers can simply continue invoking the standard POSIX APIs (`getpwnam()`, `getgrnam()`, etc) and a Host Account Management NSS module will ensure that the credentials get retrieved from hamd running on the host. The HAM NSS module (`libnss_ham.so.2`) need be installed and configured (`/etc/nsswitch.conf`) in the containers that require access to the host's Linux database.
+
+### 1.1.2 Configuration and Management Requirements
+An interface and accompanying CLI must be developed for local user management. Local users should be configurable like any other feature: via CLI, REST, and gNMI. Additionally, users may also be created and managed via Linux commands in the Bash shell. This will add additional complexity and require a service to sync between the Redis DB and the Linux user database.
+
+
+### 1.1.3 Scalability Requirements
+
+#### 1.1.3.1 Translib
+- Translib will cache all the user information along with the privilege and resource information to avoid the overhead of querying them every time we receive a request.
+- Will rely on notification to update any change in the user information, privilege or resource information
+
+### 1.1.4 Warm Boot Requirements
+N/A
+
+## 1.2 Design Overview
+### 1.2.1 Basic Approach
+
+The Translib code (also in sonic-mgmt-framework) will be modified to support RBAC via Roles, rather than Groups. It will receive username data from the REST/gNMI NBIs and perform the role lookup for a given user.
```

Comment:
We should be clear that the role info will now be stored in the User / Role Tables, rather than looked up via the POSIX APIs. We also need to see what is the performance impact of these DB lookups and call this out as a risk in the HLD.